examples/conversation_with_agent_with_finetuned_model/huggingface_model.py

```diff
@@ -0,0 +1,589 @@
+            local_model_path (str, optional): Local file path to a
+                                              pre-trained model
+                                              and its tokenizer.
+            fine_tune_config (dict, optional): Configuration for
+                                               fine-tuning the model.
+            **kwargs: Additional keyword arguments.
+        """
+        super().__init__(config_name=config_name)
+        self.model = None
+        self.max_length = max_length  # Set max_length as an attribute
+        self.pretrained_model_name_or_path = pretrained_model_name_or_path
+        script_path = os.path.abspath(__file__)
+        script_dir = os.path.dirname(script_path)
+        dotenv_path = os.path.join(script_dir, ".env")
+        _ = load_dotenv(dotenv_path)  # read local .env file
+        self.huggingface_token = os.getenv("HUGGINGFACE_TOKEN")
+
+        if device is None:
+            self.device_map = "auto"
+            self.device = torch.device(
+                "cuda" if torch.cuda.is_available() else "cpu",
+            )
+        else:
+            self.device_map = device
+            self.device = device
+
+        self.load_model(
+            pretrained_model_name_or_path,
+            local_model_path=local_model_path,
+            fine_tune_config=fine_tune_config,
+        )
+        self.load_tokenizer(
+            pretrained_model_name_or_path,
+            local_model_path=local_model_path,
+        )
+
+        if data_path is not None:
+            self.model = self.fine_tune_training(
+                self.model,
+                self.tokenizer,
+                data_path,
+                output_dir,
+                token=self.huggingface_token,
+                fine_tune_config=fine_tune_config,
+            )
+
+    def __call__(
+        self,
+        inputs: List[Dict[str, Any]],
+        **kwargs: Any,
+    ) -> ModelResponse:
+        """Process the input data to generate a response from the model.
+
+        This method tokenizes the input text, generates
+        a response using the model,
+        and then decodes the generated tokens into a string.
+
+        Arguments:
+            input (list): A list of dictionaries where each dictionary contains
+                          'name', 'role' and 'content' keys
+                          and their respective values.
+            **kwargs: Additional keyword arguments for the
+                      model's generate function.
+
+        Returns:
+            ModelResponse: An object containing the generated
+                           text and raw model output.
+
+        Raises:
+            Exception: If an error occurs during text generation.
+        """
+
+        try:
+            # Tokenize the input text
+            concatenated_input = "\n ".join(
+                [f"{d.get('role')}: {d['content']}" for d in inputs],
+            )
+            input_ids = self.tokenizer.encode(
```

Comment:
Sorry could you elaborate on this or provide an example? This system prompt is defined in `conversation_with_agent_with_finetuned_model.py`.

Reply:
e.g. using the API `tokenizer.apply_chat_template` in huggingface transformers.

Reply:
I have been tinkering with `apply_chat_template` for the past week, and here is what I found:
1. Pre-trained models do not have special tokens to indicate the beginning and the end of the assistant/user dialogue. Therefore, in order to use `apply_chat_template` (https://github.com/huggingface/trl/blob/9e9dc96e676a3601882b5cf11842bd22267fd2c5/trl/models/utils.py#L65), there is a need to add special tokens for instruction fine-tuning purpose. 
2. `setup_chat_format` is necessary for adding special tokens, because token embeddings need to be resized () the existing implementation of `setup_chat_format` is problematic (see https://github.com/huggingface/trl/issues/1412), so I tried to implement it myself.
3. Both my version and the official version of `setup_chat_format` do not seem to have good synergy with PEFT fine-tuning. The reason is, PEFT fine-tuning does not fine-tune the token embedding layer, so the token embeddings of the newly added tokens do not get optimized. This is likely the reason of an issue I could not resolve: the fine-tuned model works as expected before saving, but after saving and loading it outputs gibberish.

Reply:
Update: regarding point 3, the issue has been identified: resizing input token embeddings (https://github.com/huggingface/trl/blob/9e9dc96e676a3601882b5cf11842bd22267fd2c5/trl/models/utils.py#L98C5-L100C6)initializes the token embeddings of the newly introduced tokens randomly, and since saving PEFT model does not save its input embedding layer, during inference this resizing needs to be performed again, which leads to different initializations for the new tokens compared to the values used during fine-tuning. The solution is to save and load the input embedding layer together with the PEFT trainable parameters.

Reply:
Update: implemented a solution such that:
1. If loading a model from huggingface, check if it has a chat_template. If yes use the existing chat template; if no, use the version of `setup_chat_format` I implemented.
2. If loading a model from local path, check if it has a chat_template. If yes, check if the embedding layer corresponds to the same number of tokens as the tokenizer; if no, resize the embedding layer and use the existing chat template. If no existing chat template, use the version of `setup_chat_format` I implemented.